lib/src/common/disposable.dart

```diff
@@ -21,25 +21,45 @@ import 'package:meta/meta.dart';
 import 'package:w_common/src/common/disposable_manager.dart';
 import 'package:w_common/src/common/managed_stream_subscription.dart';
 
-// ignore: one_member_abstracts
-abstract class _Disposable {
-  Future<Null> dispose();
-}
-
-class InternalDisposable implements _Disposable {
+class InternalDisposable implements SimpleDisposable {
   Disposer _disposer;
+  final Completer<Null> _didDispose = new Completer<Null>();
+  bool _isDisposing = false;
 
   InternalDisposable(this._disposer);
 
   @override
   Future<Null> dispose() {
-    var disposeFuture = _disposer != null ? _disposer() : null;
-    _disposer = null;
-    if (disposeFuture == null) {
-      return new Future.value();
+    if (isDisposed) {
+      return null;
     }
-    return disposeFuture.then((_) => null);
+    if (_isDisposing) {
+      return didDispose;
+    }
```

Comment:
🔮 Could this could be simplified? I wonder if `didDispose` future already resolved would a return of its value return on the current or next future?

```dart
if (isDisposedOrDisposing) {
    return didDispose;
}
```